File: reader/server/core/backend/src/main/java/com/richodemus/reader/backend/feed/FeedServicePort.java
package com.richodemus.reader.backend.feed;

import com.richodemus.reader.backend.model.Feed;
import com.richodemus.reader.dto.FeedId;
import com.richodemus.reader.dto.FeedUrl;
import com.richodemus.reader.youtube_feed_service.Video;
import com.richodemus.reader.youtube_feed_service.YoutubeFeedService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

import static java.util.stream.Collectors.toList;

@Component
public class FeedServicePort implements FeedRepository {
    private final YoutubeFeedService feedService;

    @Autowired
    public FeedServicePort(YoutubeFeedService feedService) {
        this.feedService = feedService;
    }

    @Override
    public Optional<Feed> getFeed(FeedId feedId) {
        var channel = feedService.getChannel(feedId);
        List<Video> videos = feedService.getVideos(feedId);
        if (videos.isEmpty() || channel == null)
            return Optional.empty();
        return Optional.of(new Feed(feedId, channel.getName(), convertVideos(videos)));
    }

    @Override
    public FeedId getFeedId(FeedUrl feedUrl) {
        return feedService.getFeedId(feedUrl);
    }

    private List<com.richodemus.reader.backend.model.Item> convertVideos(List<Video> videos) {
        return videos.stream()
                .map(this::convertVideo)
                .collect(toList());
    }

    private com.richodemus.reader.backend.model.Item convertVideo(Video video) {
        return new com.richodemus.reader.backend.model.Item(video.getId(), video.getTitle(), video.getDescription(), video.getUploadDate().toString(), "https://youtube.com/watch?v=" + video.getId(), video.getDuration(), video.getViews());
    }
}

File: reader/server/core/backend/src/main/kotlin/com/richodemus/reader/backend/feed/FeedServicePort.kt
package com.richodemus.reader.backend.feed

import com.richodemus.reader.backend.model.Feed
import com.richodemus.reader.backend.model.Item
import com.richodemus.reader.dto.FeedId
import com.richodemus.reader.dto.FeedUrl
import com.richodemus.reader.youtube_feed_service.Video
import com.richodemus.reader.youtube_feed_service.YoutubeFeedService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.stream.Collectors

@Component
class FeedServicePort @Autowired constructor(private val feedService: YoutubeFeedService) : FeedRepository {
    override fun getFeed(feedId: FeedId): Feed? {
        val channel = feedService.getChannel(feedId)
        val videos = feedService.getVideos(feedId)
        return if (videos.isEmpty() || channel == null) null else (Feed(feedId, channel.name, convertVideos(videos)))
    }

    override fun getFeedId(feedUrl: FeedUrl): FeedId? {
        return feedService.getFeedId(feedUrl)
    }

    private fun convertVideos(videos: List<Video>): List<Item> {
        return videos.stream()
                .map { video: Video -> convertVideo(video) }
                .collect(Collectors.toList())
    }

    private fun convertVideo(video: Video): Item {
        return Item(video.id, video.title, video.description, video.uploadDate.toString(), "https://youtube.com/watch?v=" + video.id, video.duration, video.views)
    }
}
